keel-core/src/test/kotlin/com/netflix/spinnaker/keel/services/ApplicationServiceTests.kt

```diff
@@ -127,60 +121,111 @@ class ApplicationServiceTests : JUnit5Minutests {
     }
 
     before {
-      clock.reset()
+      every { repository.getDeliveryConfigForApplication(application) } returns deliveryConfig
     }
 
-    after {
-      clearAllMocks()
+    context("artifact summaries by application") {
+      before {
+        // repository.artifactVersions(artifact)
+        every { repository.artifactVersions(artifact) } returns listOf(version0, version1, version2, version3, version4)
+      }
+
+      context("all versions are pending in all environments") {
+        before {
+          every {
+            repository.getEnvironmentSummaries(deliveryConfig)
+          } returns deliveryConfig.environments.map { env ->
+            toEnvironmentSummary(env)
+          }
+
+          every {
+            repository.constraintStateFor(deliveryConfig.name, any(), any<String>())
+          } returns emptyList()
+
+          every {
+            dependsOnEvaluator.canPromote(artifact, any(), deliveryConfig, any())
+          } returns true
+        }
+
+        test("artifact summary shows all versions pending in all environments") {
+          val summaries = applicationService.getArtifactSummariesFor(application)
+
+          expectThat(summaries) {
+            hasSize(1)
+            first().and { // TODO: replace with withFirst when Strikt > 0.26.0 hits
+              get { name }.isEqualTo(artifact.name)
+              with(ArtifactSummary::versions) {
+                hasSize(versions.size)
+                all {
+                  with(ArtifactVersionSummary::environments) {
+                    hasSize(environments.size)
+                    all {
+                      get { state }.isEqualTo(PENDING.name.toLowerCase())
+                    }
+                  }
+                }
+              }
```

Comment:
This looks pretty good, @robfletcher! Thanks for taking the time to lend your expertise to improving these tests. My only feedback is that I find the `with` blocks a bit hard to read above, and wondered if there was a way to rewrite that in a simpler way. For example, would it work to replace `with(ArtifactSummary::versions) {` with `get { versions }.and {`, or even `with(this.versions) {`, or is `with` a strikt function here?